samza-azure/src/main/java/org/apache/samza/coordinator/AzureJobCoordinator.java

```diff
@@ -443,18 +440,24 @@ private void onNewJobModelAvailable(final String nextJMVersion) {
   private void onNewJobModelConfirmed(final String nextJMVersion) {
     LOG.info("pid=" + processorId + "new version " + nextJMVersion + " of the job model got confirmed");
 
-    // Delete previous value
-    if (table.getEntity(currentJMVersion.get(), processorId) != null) {
-      table.deleteProcessorEntity(currentJMVersion.get(), processorId);
-    }
-    if (table.getEntity(INITIAL_STATE, processorId) != null) {
-      table.deleteProcessorEntity(INITIAL_STATE, processorId);
-    }
+    String prevVersion = currentJMVersion.get();
 
-    //Start heartbeating to new entry only when barrier reached.
+    //Start heart-beating to new entry only when barrier reached.
     //Changing the current job model version enables that since we are heartbeating to a row identified by the current job model version.
     currentJMVersion.getAndSet(nextJMVersion);
 
+    // Delete previous value
```

Comment:
Not directly related to these changes, but it would be nicer if this functionality was abstracted into a separate class (AzureUtils or something) instead of bleeding out all these details. Something like:
azureUtil.cleanupVersion(pervVersion()
azureUtil.cleanupInitialState() and so on.

Reply:
Yeah I agree. I didn't want to make more refactoring in this patch and just focus on the bug here. I am working on another patch with refactoring and code cleanup. There is a lot of it!